File: test-prompt.py
#!/usr/bin/env -S uv run --quiet --script
# /// script
# requires-python = ">=3.12"
# dependencies = []
# ///

"""
Test script to generate and display the prompt for a specific PR without running bob.
Usage: ./test-prompt.py <repo_path> <pr_number>
"""

import os
import sys
import subprocess
from pathlib import Path

# Import functions from pr-loop.py by executing it in the same namespace
exec(Path("pr-loop.py").read_text())


def main():
    if len(sys.argv) != 3:
        print("Usage: ./test-prompt.py <repo_path> <pr_number>", file=sys.stderr)
        print("\nExamples:", file=sys.stderr)
        print("  ./test-prompt.py /path/to/repo 123", file=sys.stderr)
        print("  ./test-prompt.py . 456", file=sys.stderr)
        sys.exit(1)

    repo_path = Path(sys.argv[1]).resolve()

    # Validate repository
    if not validate_repository(repo_path):
        sys.exit(1)

    # Change to repository directory
    os.chdir(repo_path)

    try:
        pr_number = int(sys.argv[2])
    except ValueError:
        print(f"Error: '{sys.argv[2]}' is not a valid PR number", file=sys.stderr)
        sys.exit(1)

    print(f"Gathering context for PR #{pr_number}...\n", file=sys.stderr)

    # Get PR basic info
    returncode, stdout, stderr = run_command([
        "gh", "pr", "view", str(pr_number),
        "--json", "number,title,headRefName,baseRefName,url"
    ])

    if returncode != 0:
        print(f"Error fetching PR #{pr_number}: {stderr}", file=sys.stderr)
        sys.exit(1)

    import json
    pr_info = json.loads(stdout)

    # Get guidelines and commit examples
    guidelines = get_pr_guidelines()
    commit_examples = get_commit_history_examples() if not guidelines else ""

    # Gather context
    head_branch = pr_info["headRefName"]
    base_branch = pr_info.get("baseRefName", "main")
    url = pr_info["url"]

    pr_details, pr_context = gather_pr_context(pr_number, head_branch, base_branch, url)

    # Build prompt
    prompt = build_pr_prompt(pr_details, pr_context, guidelines, commit_examples)

    # Print summary to stderr
    print("\n" + "="*80, file=sys.stderr)
    print("PROMPT GENERATION SUMMARY", file=sys.stderr)
    print("="*80, file=sys.stderr)
    print(f"PR: #{pr_number} - {pr_info['title']}", file=sys.stderr)
    print(f"Branch: {head_branch} → {base_branch}", file=sys.stderr)
    print(f"Prompt size: {len(prompt)} characters", file=sys.stderr)
    print(f"Comments: {len(pr_context['comments'])}", file=sys.stderr)
    print(f"Review comments: {len(pr_context['review_comments'])}", file=sys.stderr)
    print(f"Commits: {len(pr_context['commits'])}", file=sys.stderr)
    print(f"Files changed: {len(pr_context['files'])}", file=sys.stderr)
    print(f"Has conflicts: {pr_context['conflicts'].get('has_conflicts', False)}", file=sys.stderr)
    print(f"CI checks: {pr_context['ci_status'].get('total_checks', 0)}", file=sys.stderr)
    print(f"Failed checks: {pr_context['ci_status'].get('failed', 0)}", file=sys.stderr)
    print("="*80, file=sys.stderr)
    print("\nGenerated prompt (stdout):\n", file=sys.stderr)

    # Print the actual prompt to stdout so it can be piped
    print(prompt)


if __name__ == "__main__":
    main()

File: test-prompt.ts
#!/usr/bin/env node
/**
 * Test script to generate and display the prompt for a specific PR without running pi.
 *
 * Ported from test-prompt.py. Usage: ./test-prompt.ts <repo_path> <pr_number>
 */

import { chdir } from "node:process";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";
import { normalizeCiStatusCounts } from "./ci_status_model";
import { hasActiveMergeConflicts } from "./conflict_model";
import { prContextCiStatus, prContextCommits, prContextConflicts } from "./pr_context_access_model";
import { prContextTelemetrySummary } from "./pr_context_log_model";
import { prDetailsBaseBranch, prDetailsHeadBranch, prDetailsTitle, prDetailsUrl } from "./pr_details_access_model";
import {
  validateRepository,
  getPrGuidelines,
  getCommitHistoryExamples,
  gatherPrContext,
  buildPrPrompt,
} from "./pr-loop";
import { initializeTelemetry, recordPromptRendered, shutdownTelemetry } from "./telemetry";
import { ExecutionPolicy } from "./execution_policy";

function runCommand(
  cmd: string[],
): { returncode: number; stdout: string; stderr: string } {
  const result = new ExecutionPolicy().runCommandSync(cmd[0] ?? "", cmd.slice(1), {
    maxBuffer: 10 * 1024 * 1024,
  });
  return { returncode: result.status, stdout: result.stdout, stderr: result.stderr };
}

async function main(): Promise<void> {
  const repoPathArg = process.argv[2];
  const prArg = process.argv[3];

  if (repoPathArg === undefined || prArg === undefined) {
    console.error("Usage: ./test-prompt.ts <repo_path> <pr_number>");
    console.error("\nExamples:");
    console.error("  ./test-prompt.ts /path/to/repo 123");
    console.error("  ./test-prompt.ts . 456");
    process.exit(1);
  }

  const repoPath = resolve(repoPathArg);

  if (!validateRepository(repoPath)) {
    process.exit(1);
  }

  chdir(repoPath);

  if (!/^[+-]?\d+$/.test(prArg.trim())) {
    console.error(`Error: '${prArg}' is not a valid PR number`);
    process.exit(1);
  }
  const prNumber = Number.parseInt(prArg, 10);

  console.error(`Gathering context for PR #${prNumber}...\n`);

  const { returncode, stdout, stderr } = runCommand([
    "gh",
    "pr",
    "view",
    String(prNumber),
    "--json",
    "number,title,headRefName,baseRefName,url",
  ]);

  if (returncode !== 0) {
    console.error(`Error fetching PR #${prNumber}: ${stderr}`);
    process.exit(1);
  }

  const prInfo = JSON.parse(stdout) as Record<string, unknown>;

  const title = prDetailsTitle(prInfo);
  const headBranch = prDetailsHeadBranch(prInfo);
  const baseBranch = prDetailsBaseBranch(prInfo);
  const url = prDetailsUrl(prInfo);

  const guidelines = getPrGuidelines();
  const commitExamples = guidelines ? "" : getCommitHistoryExamples(baseBranch);

  const [prDetails, prContext] = await gatherPrContext(prNumber, headBranch, baseBranch, url);

  const prompt = buildPrPrompt(prDetails, prContext, guidelines, commitExamples);
  recordPromptRendered("test_prompt.pr", prompt, {
    "merge_god.pr_number": prNumber,
  });

  const contextSummary = prContextTelemetrySummary(prContext);
  const commits = prContextCommits(prContext);
  const hasConflicts = hasActiveMergeConflicts(prContextConflicts(prContext));
  const ciCounts = normalizeCiStatusCounts(prContextCiStatus(prContext));

  const sep = "=".repeat(80);
  console.error("\n" + sep);
  console.error("PROMPT GENERATION SUMMARY");
  console.error(sep);
  console.error(`PR: #${prNumber} - ${title}`);
  console.error(`Branch: ${headBranch} → ${baseBranch}`);
  console.error(`Prompt size: ${prompt.length} characters`);
  console.error(`Comments: ${contextSummary.comment_count}`);
  console.error(`Review comments: ${contextSummary.review_comment_count}`);
  console.error(`Commits: ${commits.length}`);
  console.error(`Files changed: ${contextSummary.file_count}`);
  console.error(`Has conflicts: ${hasConflicts}`);
  console.error(`CI checks: ${ciCounts.total}`);
  console.error(`Failed checks: ${ciCounts.failed}`);
  console.error(sep);
  console.error("\nGenerated prompt (stdout):\n");

  console.log(prompt);
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? "").href) {
  initializeTelemetry();
  main()
    .then(() => {
      void shutdownTelemetry().finally(() => process.exit(0));
    })
    .catch((e) => {
      console.error(e);
      void shutdownTelemetry().finally(() => process.exit(1));
    });
}
